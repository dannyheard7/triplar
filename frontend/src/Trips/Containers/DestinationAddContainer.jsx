import React from "react";
import DestinationForm from "Trips/Components/DestinationForm";
import api from "utils/api.js"
import FormContainer from "Forms/Containers/FormContainer";

export default class DestinationAddContainer extends React.Component {
    constructor(props) {
        super(props);

        this.state = {
            showDestinationAdd: false
        };

        this.onClick = this.onClick.bind(this);
        this.onSuccess = this.onSuccess.bind(this)
    }

    onClick() {
        this.setState({showDestinationAdd: !this.state.showDestinationAdd});
    }

    onSuccess(itinerary) {
        this.setState({showDestinationAdd: false});
        this.props.onSuccess(itinerary)
    }

    apiFunc = api.curryAPIFunc(api.addDestinationToTrip, this.props.tripId);

    render () {
        return (
            <div>
                <div className="row">
                    <div className="col md-2">
                        <button className="btn btn-primary" onClick={this.onClick}>Add Destination</button>
                    </div>
                </div>
                {this.state.showDestinationAdd &&
                    <div className="card">
                        <FormContainer onSuccess={this.onSuccess} apiFunction={this.apiFunc}>
                            <DestinationForm submitLabel="Add"/>
                        </FormContainer>
                    </div>
                }
            </div>
        )
    }
}