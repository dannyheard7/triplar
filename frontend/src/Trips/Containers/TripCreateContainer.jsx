import React from "react";
import api from "utils/api.js"
import {withRouter} from "react-router-dom";

import TripForm from "Trips/Components/TripForm";
import FormContainer from "../../Forms/Containers/FormContainer";

export class TripCreateContainer extends React.Component {
    constructor(props) {
        super(props);

        this.onSuccess = this.onSuccess.bind(this);
        this.onClick = this.onClick.bind(this);

        this.state = {
            showTripCreate: false
        };
    }

    onSuccess(trip) {
        this.setState({showTripCreate: false});
        this.props.history.push('/trips/' + trip.id)
    };

    onClick() {
        this.setState({showTripCreate: !this.state.showTripCreate});
    }

    render () {
        return (
            <div>
                <div className="row">
                    <div className="col md-2">
                        <button className="btn btn-primary" onClick={this.onClick}>New Trip</button>
                    </div>
                </div>
                {this.state.showTripCreate &&
                    <div className="card">
                        <FormContainer onSuccess={this.onSuccess} apiFunction={api.createTrip}>
                            <TripForm submitLabel="Create" />
                        </FormContainer>
                    </div>
                }
            </div>
        )
    }
}

export default withRouter(TripCreateContainer)