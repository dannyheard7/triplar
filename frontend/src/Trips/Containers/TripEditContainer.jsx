import React from "react";
import api from "utils/api.js";

import TripForm from "Trips/Components/TripForm.jsx";
import FormContainer from "../../Forms/Containers/FormContainer";

export default class TripEditContainer extends React.Component {
    constructor(props) {
        super(props);

        this.onSuccess = this.onSuccess.bind(this);
    }

    onSuccess(trip) {
        this.props.onUpdate(trip);
    };

    apiFunc = api.curryAPIFunc(api.patchTrip, this.props.trip.id);

    render() {
        return (
            <div className="card">
                <FormContainer onSuccess={this.onSuccess} apiFunction={this.apiFunc} >
                    <TripForm submitLabel="Save" trip={this.props.trip}/>
                </FormContainer>
            </div>
        );
    }
}

TripEditContainer.defaultProps = {
    trip: {}
};