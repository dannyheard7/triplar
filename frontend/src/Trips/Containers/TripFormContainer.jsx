import React from "react";
import TripForm from "Trips/Components/TripForm";

export default class TripFormContainer extends React.Component {
    constructor(props) {
        super(props);

        this.state = {
            trip: this.props.trip,
            errors: []
        };

        this.handleChange = this.handleChange.bind(this);
        this.handleSubmit = this.handleSubmit.bind(this);
    }


    handleChange(event) {
        let trip = this.state.trip;
        trip[event.target.name] = event.target.value;

        this.setState({trip: trip});
    }

    handleSubmit(e) {
        e.preventDefault();

        this.props.apiFunction(this.state.trip).then(response => {
            if(response.status === 400) {
                this.setState({errors: response.data})
            } else {
                this.setState({trip: {}, errors: []});
                this.props.onSuccess(response.data);
            }
        });
    }

    render () {
        return (
            <TripForm onChange={this.handleChange} onSubmit={this.handleSubmit} submitLabel={this.props.submitLabel}
                      errors={this.state.errors} suggestions={this.state.suggestions} trip={this.state.trip} />
        )
    }

}

TripFormContainer.defaultProps = {
    trip: {}
};