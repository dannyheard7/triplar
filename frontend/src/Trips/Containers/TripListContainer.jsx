import React from "react";
import api from "utils/api.js";

import TripList from "Trips/Components/TripList";

export default class TripListContainer extends React.Component {
    constructor(props) {
        super(props);

        this.state = {
            trips: []
        };
    }

    componentDidMount() {
        this.getTrips();
    }

    componentWillReceiveProps(props) {
        if(props.update) {
            this.getTrips();
        }
    }

    getTrips() {
        api.getTrips().then(res => {
            this.setState({trips: res.data});
        });
    }

    render() {
        return (<TripList trips={this.state.trips}/>);
    }
}