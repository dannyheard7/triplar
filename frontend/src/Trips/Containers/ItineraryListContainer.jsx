import React from "react";
import DestinationItinerary from "../Components/DestinationItinerary";
import DestinationAddContainer from "Trips/Containers/DestinationAddContainer";

export default class ItineraryListContainer extends React.Component {
     constructor(props) {
        super(props);

        this.state = {
            itineraries: this.props.itineraries
        };

        this.onDestinationAdd = this.onDestinationAdd.bind(this);
    }

    onDestinationAdd(destination) {
         this.setState({itineraries: this.state.itineraries.concat([destination])})
    }

    render() {
        const itineraries = this.state.itineraries.map((itinerary) =>
            <DestinationItinerary key={itinerary.id} itinerary={itinerary}/>
        );

        return (
            <div>
                {itineraries}
                <DestinationAddContainer tripId={this.props.tripId} onSuccess={this.onDestinationAdd}/>
            </div>
        );
    }
}