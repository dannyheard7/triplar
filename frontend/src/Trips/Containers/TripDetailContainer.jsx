import React from "react";
import axios from "axios";

import TripDetail from "Trips/Components/TripDetail";
import TripManagementRow from "Trips/Components/TripManagementRow";
import TripDelete from "Trips/Components/TripDelete";
import TripEditContainer from "Trips/Containers/TripEditContainer";

export default class TripDetailContainer extends React.Component {
     constructor(props) {
        super(props);

        this.state = {
            trip: null,
            edit: false,
            delete: false
        };

        this.onDeleteClick = this.onDeleteClick.bind(this);
        this.onEditClick = this.onEditClick.bind(this);
        this.onUpdate = this.onUpdate.bind(this);
    }

    componentDidMount() {
        let url = "/trips/" + this.props.match.params.id;

        axios.get(url).then(res => {
            const trip = res.data;
            this.setState({ trip });
        });
    }

    onUpdate(trip) {
        this.setState({trip: trip, edit: false});
    }

    onEditClick(e) {
        this.setState({edit: true});
    }

    onDeleteClick(e) {
        this.setState({delete: true});
    }

    render() {
        const trip = this.state.trip;

        if (!trip) {
            return <div>Loading</div>
        }
        if (this.state.edit) {
            return <TripEditContainer trip={trip} onUpdate={this.onUpdate}/>
        }
        return (
            <div>
                <TripDetail trip={trip} />
                <TripManagementRow trip={trip} onEditClick={this.onEditClick} onDeleteClick={this.onDeleteClick} />
                {this.state.delete && <TripDelete trip={trip} router={this.props.router}/>}
            </div>
        );
    }
}