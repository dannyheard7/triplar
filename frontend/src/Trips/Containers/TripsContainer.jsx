import React from "react";

import TripListContainer from "Trips/Containers/TripListContainer";
import TripCreateContainer from "Trips/Containers/TripCreateContainer";

// TODO: Move into a functional component

export default class TripsContainer extends React.Component {
     render() {
        return (
            <div>
                <TripListContainer />
                <TripCreateContainer />
            </div>
        )
    }
}