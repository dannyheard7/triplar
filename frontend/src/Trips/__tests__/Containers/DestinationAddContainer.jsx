import React from "react";
import {shallow} from "enzyme";

import ShallowRenderer from 'react-test-renderer/shallow';
const renderer = new ShallowRenderer();
const faker = require('faker');

import { TripCreateContainer } from "Trips/Containers/TripCreateContainer";
import FormContainer from "Forms/Containers/FormContainer";
import DestinationAddContainer from "Trips/Containers/DestinationAddContainer";

jest.mock('utils/api.js');

const api = require('utils/api.js');

describe('<DestinationAddContainer />', () => {
    const event = {
        preventDefault: () => {},
        target: {
            value: "",
            name: "",
        },
    };

    const itinerary = {'id': faker.random.number(), 'city': {'name_std': faker.address.city(), 'country': faker.address.country()}};


    test('renders correctly', () => {
        const result = renderer.render(<DestinationAddContainer tripId="1" />);
        expect(result).toMatchSnapshot();
    });

    test('calls onSuccess after receiving success from <FormContainer />', (function () {
        const spy = jest.spyOn(DestinationAddContainer.prototype, 'onSuccess');
        const stub = jest.fn();
        const container = shallow(<DestinationAddContainer  tripId={faker.random.number()} onSuccess={stub} />);
        container.setState({showDestinationAdd: true});
        container.find(FormContainer).prop('onSuccess')(itinerary);

        expect(spy).toBeCalledWith(itinerary);
        expect(stub).toBeCalledWith(itinerary)
    }));

    test('addDestinationToTrip called with correct arguments', () => {
        const spy = jest.spyOn(api.default, 'addDestinationToTrip');
        const id = faker.random.number();

        const container = shallow(<DestinationAddContainer  tripId={id} />);
        container.instance().apiFunc(itinerary);

        expect(spy).toBeCalledWith(id, itinerary);

        spy.mockRestore();
    });


    describe('destination create visible', () => {
        let container = null;

        beforeEach(function () {
            container = shallow(<DestinationAddContainer  tripId={faker.random.number()} onSuccess={jest.fn()}/>);
            container.setState({showDestinationAdd: true});
        });

        test('renders <FormContainer /> component when showTripCreate is true', () => {
            expect(container.find(FormContainer).length).toEqual(1);
        });

        test('updates state.showDestinationAdd when create button clicked', () => {
            container.find('.btn').simulate('click');
            expect(container.state('showDestinationAdd')).toBeFalsy();
        });


        test('does not render <FormContainer /> after a trip has been created', () => {
            expect(container.find(FormContainer).length).toEqual(1);
            container.find(FormContainer).prop('onSuccess')(itinerary);
            container.update();

            expect(container.state('showDestinationAdd')).toEqual(false);
            expect(container.find(FormContainer).length).toEqual(0);
        });

    });

});