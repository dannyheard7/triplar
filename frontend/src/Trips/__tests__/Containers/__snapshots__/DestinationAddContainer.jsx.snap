// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<DestinationAddContainer /> renders correctly 1`] = `
<div>
  <div
    className="row"
  >
    <div
      className="col md-2"
    >
      <button
        className="btn btn-primary"
        onClick={[Function]}
      >
        Add Destination
      </button>
    </div>
  </div>
</div>
`;
